pkg/batchscheduler/scheduler_manager.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2019 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package batchscheduler
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/client-go/rest"
+
+	"github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/batchscheduler/interface"
+	"github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/batchscheduler/volcano"
+)
+
+type SchedulerManager interface {
+	GetScheduler(schedulerName string) (schedulerinterface.BatchScheduler, error)
+}
+
+type schedulerInitializeFunc func(config *rest.Config) (schedulerinterface.BatchScheduler, error)
+
+var schedulerContainers = map[string]schedulerInitializeFunc{
+	volcano.GetPluginName(): volcano.New,
+}
+
+func GetRegisteredNames() []string {
+	var pluginNames []string
+	for key := range schedulerContainers {
+		pluginNames = append(pluginNames, key)
+	}
+	return pluginNames
+}
+
+type SchedulerManagerImpl struct {
+	sync.Mutex
+	config  *rest.Config
+	plugins map[string]schedulerinterface.BatchScheduler
+}
+
+func NewSchedulerManager(config *rest.Config) SchedulerManager {
+	manager := SchedulerManagerImpl{
+		config:  config,
+		plugins: make(map[string]schedulerinterface.BatchScheduler),
+	}
+	return &manager
+}
+
+func (batch *SchedulerManagerImpl) GetScheduler(schedulerName string) (schedulerinterface.BatchScheduler, error) {
+	iniFunc, registered := schedulerContainers[schedulerName]
+	if !registered {
+		return nil, fmt.Errorf("unregistered scheduler plugin %s", schedulerName)
+	}
+
+	batch.Lock()
+	defer batch.Unlock()
+
+	if plugin, existed := batch.plugins[schedulerName]; existed && plugin != nil {
+		return plugin, nil
+	} else if existed && plugin == nil {
+		return nil, fmt.Errorf(
+			"failed to get scheduler plugin %s, previous initialization has failed", schedulerName)
+	} else {
+		if plugin, err := iniFunc(batch.config); err != nil {
```

Comment:
I don't think we need to add an entry in this case. We can use the absence of a scheduler as the signal instead.

Reply:
There are three different cases.
1. scheduler exists. //Normal case
2. scheduler exists but it's none. //Abnormal case. We have this scheduler supported, but failed to initialize before.
3. scheduler not exists, //Normal case, we will initialize this now.
Which one are you saying we don't need to consider about?

Reply:
I meant we don't really treat the case that the scheduler failed to be initialized differently than the case that the scheduler simply is not registered. In both cases, the scheduler doesn't exist in the registry.

Reply:
Just want to save the time of creating scheduler repeatly.

Reply:
OK, that makes sense.